README.md
# hiwin README

hiwin is a vscode extension to manage vscode window on ms-windows platform, it offers the control of opacity, size, position, z-mode, max/minimize about window.

auto set windows properties when the vscode window is open.
    "hiWin.auto-load": true

``` json
{
    "hinWin.properties.opacity": {
        "type": "integer",
        "minimum": 180,
        "maximum": 255
    },
    "hinWin.properties.z-order": {
        "type": "string",
        "enum": [
        "bottom",
        "no top most",
        "top",
        "top most"
        ],
        "enumDescriptions": [
        "Place vscode window at bottom",
        "Place vscode window above non-topmost windows",
        "Place vscode window above all",
        "Set vscode window as top most layer"
        ]
    },
    "hinWin.properties.show": {
        "type": "string",
        "enum": [
        "hide",
        "maximize",
        "minimize",
        "activate"
        ],
        "enumDescriptions": [
        "Hide the vscode window",
        "Maximize the vscode window",
        "Minimize the vscode window",
        "Activate the vscode window"
        ]
    },
    "hinWin.properties.pos-and-size": {
        "type": [
        "array",
        "null"
        ],
        "properties": {
        "type": "integer"
        },
        "minItems": 4,
        "maxItems": 4,
        "default": null
    }
}
```

for example
``` json
{
    "hiWin.properties": {
        "hinWin.properties.opacity": 220,
        "hinWin.properties.z-order": "top most",
        "hinWin.properties.show": "activate",
        "hinWin.properties.pos-and-size": [
            0,
            0,
            750,
            500
        ]
    }

}
```

using command `hiWin: Update Window properties` to update window


CHANGELOG.md
# Change Log

All notable changes to the "hiwin" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

Win.cs
using System;
using System.Runtime.InteropServices;
using System.Diagnostics;

using Windows;

namespace Hi
{
    public static class Win
    {
        public static bool SetProperties(int pid, byte alpha, Windows.HWND z, Windows.SW show, Windows.SWP swp, int x, int y, int cx, int cy)
        {
            Process mainproc = Process.GetProcessById(pid);
            foreach (Process proc in Process.GetProcessesByName(mainproc.ProcessName))
            {
                if (proc.StartInfo.FileName != mainproc.StartInfo.FileName) { continue; }

                IntPtr hMainWnd = proc.MainWindowHandle;
                if (hMainWnd == IntPtr.Zero) { continue; }

                uint tid = User32.GetWindowThreadProcessId(hMainWnd, out pid);
                bool result = User32.EnumThreadWindows(tid, delegate(IntPtr hWnd, IntPtr lParam) {
                    if (!User32.IsWindowVisible(hWnd)) { return true; }
                    WS windowLong = User32.GetWindowLong(hWnd, GWL.EXSTYLE);
                    User32.SetWindowLong(hWnd, GWL.EXSTYLE, windowLong | WS.EX_LAYERED);

                    return User32.ShowWindow(hWnd, show) && User32.SetLayeredWindowAttributes(hWnd, 0, alpha, LWA.ALPHA) && User32.SetWindowPos(hWnd, z, x, y, cx, cy, swp);
                }, IntPtr.Zero);

                if (!result) { return false; }
            }
            return true;
        }
    }
}

namespace Windows
{
    internal static class User32
    {
        public delegate bool EnumWindowsProc(IntPtr hWnd, IntPtr lParam);

        [DllImport("user32.dll")]
        public static extern bool EnumThreadWindows(uint dwThreadId, EnumWindowsProc lpEnumFunc, IntPtr lParam);

        [DllImport("user32.dll")]
        public static extern uint GetWindowThreadProcessId(IntPtr hWnd, out int lpdwProcessId);

        [DllImport("user32.dll")]
        public static extern bool IsWindowVisible(IntPtr hWnd);

        [DllImport("user32.dll")]
        public static extern WS GetWindowLong(IntPtr hWnd, GWL nIndex);

        [DllImport("user32.dll")]
        public static extern int SetWindowLong(IntPtr hWnd, GWL nIndex, WS dwNewLong);

        [DllImport("user32.dll")]
        public static extern bool SetLayeredWindowAttributes(IntPtr hWnd, uint crKey, byte bAlpha, LWA dwFlags);

        [DllImport("user32.dll")]
        public static extern bool SetWindowPos(IntPtr hWnd, HWND hWndInsertAfter, int X, int Y, int cx, int cy, SWP uFlags);

        [DllImport("user32.dll")]
        public static extern bool  ShowWindow(IntPtr hWnd, SW nCmdShow);
    }


    public enum SW: int
    {
        HIDE = 0,
        MINIMIZE = 2,
        MAXIMIZE = 3,
        ACTIVATE = 5,
    }

    public enum HWND: int
    {
        BOTTOM = 1,
        NOTOPMOST = -2,
        TOP = 0,
        TOPMOST = -1,
    }

    public enum SWP: uint
    {
        NOMOVE = 0x0002,
        NOSIZE = 0x0001,
        ZORDER = NOSIZE | NOMOVE,
        NOZORDER = 0x0004,
        SHOWWINDOW = 0x0040,
    }

    public enum GWL: int
    {
        EXSTYLE = -20,
        HINSTANCE = -6,
        HWNDPARENT = -8,
        ID = -12,
        STYLE = -16,
        USERDATA = -21,
        WNDPROC = -4,
    }

    [Flags]
    public enum WS: int
    {
        EX_LAYERED = 0x80000,
    }

    public enum LWA: int
    {
        COLORKEY = 1,
        ALPHA = 2,
    }
}


src/extension.ts
import * as vscode from "vscode";

enum SWP {
  NOMOVE = 0x0002,
  NOSIZE = 0x0001,
  ZORDER = NOSIZE | NOMOVE,
  NOZORDER = 0x0004,
  SHOWWINDOW = 0x0040,
}

enum HWND {
  BOTTOM = 1,
  NOTOPMOST = -2,
  TOP = 0,
  TOPMOST = -1,
}

enum SW {
  HIDE = 0,
  MINIMIZE = 2,
  MAXIMIZE = 3,
  ACTIVATE = 5,
}

class HiWinProperties {
  "hinWin.properties.opacity": number;
  "hinWin.properties.z-order": string;
  "hinWin.properties.show": string;
  "hinWin.properties.pos-and-size": Array<number> | undefined;

  static convertOpacity(opacity: number = 255): number {
    return opacity;
  }

  static convertZOrder(zOrder: string = "no top most"): HWND {
    switch (zOrder) {
      case "bottom":
        return HWND.BOTTOM;
      case "top":
        return HWND.TOP;
      case "top most":
        return HWND.TOPMOST;
      case "no top most":
      default:
        return HWND.NOTOPMOST;
    }
  }

  static convertShow(show: string = "activate"): SW {
    switch (show) {
      case "hide":
        return SW.HIDE;
      case "minimize":
        return SW.MINIMIZE;
      case "maximize":
        return SW.MAXIMIZE;
      case "activate":
      default:
        return SW.ACTIVATE;
    }
  }

  static convertPosAndSize(pos_and_size: Array<number> | undefined): {
    swp: SWP;
    x: number;
    y: number;
    cx: number;
    cy: number;
  } {
    if (pos_and_size && pos_and_size.length == 4) {
      return {
        swp: SWP.SHOWWINDOW,
        x: pos_and_size[0],
        y: pos_and_size[1],
        cx: pos_and_size[2],
        cy: pos_and_size[3],
      };
    }
    return { swp: SWP.ZORDER, x: 0, y: 0, cx: 0, cy: 0 };
  }
}

class WinManagement {
  private static instance: WinManagement;

  private config: HiWinProperties | undefined;
  private constructor() {}

  public static getInstance() {
    if (!WinManagement.instance) {
      WinManagement.instance = new WinManagement();
    }
    return WinManagement.instance;
  }

  private updateConfig() {
    this.config = vscode.workspace
      .getConfiguration("hiWin")
      .get<HiWinProperties>("properties");
  }
  private getOpacity(): number {
    return HiWinProperties.convertOpacity(
      this.config?.["hinWin.properties.opacity"]
    );
  }

  private getZorder(): HWND {
    return HiWinProperties.convertZOrder(
      this.config?.["hinWin.properties.z-order"]
    );
  }

  private getShow(): SW {
    return HiWinProperties.convertShow(this.config?.["hinWin.properties.show"]);
  }

  private getPosAndSize(): {
    swp: SWP;
    x: number;
    y: number;
    cx: number;
    cy: number;
  } {
    return HiWinProperties.convertPosAndSize(
      this.config?.["hinWin.properties.pos-and-size"]
    );
  }

  public async apply(context: vscode.ExtensionContext, path: string) {
    this.updateConfig();
    if (this.config) {
      let { swp, x, y, cx, cy } = this.getPosAndSize();

      const shell = require("node-powershell");
      const ps = new shell({
        executionPolicy: "RemoteSigned",
        noProfile: true,
      });

      context.subscriptions.push(ps);

      ps.addCommand("[Console]::OutputEncoding = [Text.Encoding]::UTF8");
      ps.addCommand(`Add-Type -Path '${path}'`);
      ps.addCommand(
        `[Hi.Win]::SetProperties(${
          process.pid
        }, ${this.getOpacity()}, ${this.getZorder()}, ${this.getShow()}, ${swp}, ${x}, ${y}, ${cx}, ${cy})`
      );

      await ps.invoke();
    }
  }
}

export function activate(context: vscode.ExtensionContext) {
  const path: string = context.asAbsolutePath("./Win.cs");
  // Initialize Vs Code window
  if (vscode.workspace.getConfiguration("hiWin").get<boolean>("auto-load")) {
    WinManagement.getInstance().apply(context, path);
  } else {
    WinManagement.getInstance();
  }

  // Register command
  let disposable = vscode.commands.registerCommand(
    "hiWin.updateWin",
    async () => {
      await WinManagement.getInstance().apply(context, path);
    }
  );

  context.subscriptions.push(disposable);
}

export function deactivate() {}
